src/worker/core/team/valueChange.ts

```diff
@@ -588,4 +624,64 @@ const valueChange = async (
 	return valuesAdd - valuesRemove;
 };
 
+const getModifiedPickRank = async (
+	tid: number,
+	pidsAdd: number[],
+	pidsRemove: number[],
+) => {
+	const teamSeason = await idb.cache.teamSeasons.indexGet(
+		"teamSeasonsBySeasonTid",
+		[tid, g.get("season")],
+	);
+	const teams = g.get("numActiveTeams");
+	const record = teamSeason ? [teamSeason.won, teamSeason.lost] : [0, 0];
+	const gp = record[0] + record[1];
+	const seasonFraction = gp / g.get("numGames");
+	const players = await idb.cache.players.indexGetAll("playersByTid", tid);
+	let playerRatings = players.map(p => ({
+		pid: p.pid,
+		value: p.value,
+		ratings: {
+			ovr: p.ratings.at(-1)!.ovr,
+			ovrs: p.ratings.at(-1)!.ovrs,
+			pos: p.ratings.at(-1)!.pos,
+		},
+	}));
+	if (pidsRemove.length != 0) {
+		playerRatings = playerRatings.filter(p => !pidsRemove.includes(p.pid));
+	}
+	if (pidsAdd.length != 0) {
+		for (const pid of pidsAdd) {
+			const p = await idb.cache.players.get(pid);
+			if (p != undefined) {
+				playerRatings.push({
+					pid: p.pid,
+					value: p.value,
+					ratings: {
+						ovr: p.ratings.at(-1)!.ovr,
+						ovrs: p.ratings.at(-1)!.ovrs,
+						pos: p.ratings.at(-1)!.pos,
+					},
+				});
+			}
+		}
+	}
```

Comment:
So for this part, at first I was just doing players = players.indexGetAll and then adding the players from pidsAdd to that retrieved list directly. But then weird stuff in the trade UI was happening, the players from one team were showing up in the other team's players before any trade was executed, maybe because pushing a player onto players was changing a reference to the DB?

So then I tried making a copy of players with helpers.deepCopy, but that slowed things down a fair amount because it was copying all of a player's information, which I guess is a lot, when all this cares about is the ratings. That's why I mapped to the playerRatings right away

I'm guessing there's a cleaner way to do this, saw there was some getCopies functionality on idb, but I didn't really know how to use it, could you make this change for me or rewrite this how you think is best?

Reply:
Sure I will take a look... it is true that the player objects you get from the cache should not be modified unless you're actually trying to update the cache. But you should be able to assemble them into arrays however you want. It's only the individual player objects that matter, the array they're stored in is not meaningful.

And yeah copying is like a last resort when absolutely necessary, because it is slow.

getCopies should probably be renamed because there's now an option to not copy the values from the cache. But when it's used to look back further into the database (anything only in IndexedDB but not in the cache) then that is always a copy of what's in the database and you can mutate it without causing any problems. That's why I initially made getCopies, to merge the cache and the database, and I figured I should always deep copy the cache for consistency. But if you're careful, copying cached values is just waste.

Reply:
See 2103e7da6aeb78499c9fab61b56f5e477b609e45 it seems to work. I'll give you an opportunity to test it, and also I still need to test it in comparison to the master branch to make sure the whole PR is actually doing what it's supposed to be doing :)